src/LinuxTracing/UprobesUnwindingVisitorSampleTest.cpp

```diff
@@ -972,6 +972,417 @@ TEST_F(UprobesUnwindingVisitorSampleTest,
+                Property(&orbit_grpc_protos::FullAddressInfo::offset_in_function, 0),
+                Property(&orbit_grpc_protos::FullAddressInfo::module_name, kTargetName)),
+          AllOf(Property(&orbit_grpc_protos::FullAddressInfo::absolute_address, kTargetAddress2),
+                Property(&orbit_grpc_protos::FullAddressInfo::function_name, kFunctionName2),
+                Property(&orbit_grpc_protos::FullAddressInfo::offset_in_function, 0),
+                Property(&orbit_grpc_protos::FullAddressInfo::module_name, kTargetName)),
+          AllOf(Property(&orbit_grpc_protos::FullAddressInfo::absolute_address, kTargetAddress3),
+                Property(&orbit_grpc_protos::FullAddressInfo::function_name, kFunctionName3),
+                Property(&orbit_grpc_protos::FullAddressInfo::offset_in_function, 0),
+                Property(&orbit_grpc_protos::FullAddressInfo::module_name, kTargetName))));
+
+  EXPECT_EQ(unwinding_errors, 0);
+  EXPECT_EQ(discarded_samples_in_uretprobes_counter, 0);
+}
+
+TEST_F(UprobesUnwindingVisitorSampleTest, VisitStackSampleWillMakeUseOfLatestUserSpaceCallstack) {
+  StackSamplePerfEvent event = BuildFakeStackSamplePerfEvent();
+
+  EXPECT_CALL(return_address_manager_, PatchSample).Times(1).WillOnce(Return());
+  EXPECT_CALL(maps_, Get).Times(1).WillOnce(Return(nullptr));
+
+  std::vector<unwindstack::FrameData> libunwindstack_callstack{kFrame1, kFrame2, kFrame3};
+
+  std::vector<StackSliceView> actual_stack_slices{};
+  EXPECT_CALL(unwinder_, Unwind(event.data.pid, nullptr, _, _, _, _))
+      .Times(1)
+      .WillOnce(DoAll(SaveArg<3>(&actual_stack_slices),
+                      Return(LibunwindstackResult{
+                          libunwindstack_callstack, {}, unwindstack::ErrorCode::ERROR_NONE})));
+
+  orbit_grpc_protos::FullCallstackSample actual_callstack_sample;
+  EXPECT_CALL(listener_, OnCallstackSample).Times(1).WillOnce(SaveArg<0>(&actual_callstack_sample));
+
+  std::vector<orbit_grpc_protos::FullAddressInfo> actual_address_infos;
+  auto save_address_info =
+      [&actual_address_infos](orbit_grpc_protos::FullAddressInfo actual_address_info) {
+        actual_address_infos.push_back(std::move(actual_address_info));
+      };
+  EXPECT_CALL(listener_, OnAddressInfo).Times(3).WillRepeatedly(Invoke(save_address_info));
+
+  std::atomic<uint64_t> unwinding_errors = 0;
+  std::atomic<uint64_t> discarded_samples_in_uretprobes_counter = 0;
+  visitor_.SetUnwindErrorsAndDiscardedSamplesCounters(&unwinding_errors,
+                                                      &discarded_samples_in_uretprobes_counter);
+
+  constexpr uint64_t kUserStackSizeOld = 512;
+  constexpr uint64_t kUserStackPointerOld = 24;
+  UprobesWithStackPerfEvent user_stack_event_old{
+      .timestamp = 12,
+      .data =
+          {
+              .stream_id = 1,
+              .pid = event.data.pid,
+              .tid = event.data.tid,
+              .regs =
+                  {
+                      .abi = 10,
```

Comment:
You can use a real value for `abi`: `PERF_SAMPLE_REGS_ABI_64`

Reply:
Done.